Splatting 1 variable : AssertionError: x.args[1] in data.operators.comparison_operators

Hi,

For [`ModelPredictiveControl.jl`](https://github.com/franckgaga/ModelPredictiveControl.jl), I'm using the splatting syntax for the nonlinear programming. The code needs to be generic, that is, it needs to works with 1 or more decision variable(s).

If I do : 
```julia
using JuMP, Ipopt
optim = Model(Ipopt.Optimizer)
nvar = 1 # nvar = 1 : error, nvar = 2 : works
function Jfunc(ΔŨtup...)
    ΔŨ = collect(ΔŨtup)
    return sum(ΔŨ.^2)
end
@variable(optim, ΔŨvar[1:nvar])
register(optim, :Jfunc, nvar, Jfunc, autodiff=true)
@NLobjective(optim, Min, Jfunc(ΔŨvar...))
```
it crashes with:
```julia
ERROR: AssertionError: x.args[1] in data.operators.comparison_operators
Stacktrace:
 [1] _parse_multivariate_expression(stack::Vector{Tuple{Int64, Any}}, data::MathOptInterface.Nonlinear.Model, expr::MathOptInterface.Nonlinear.Expression, x::Expr, parent_index::Int64)
   @ MathOptInterface.Nonlinear ~/.julia/packages/MathOptInterface/BlCD1/src/Nonlinear/parse.jl:180
 [2] _parse_expression(stack::Vector{Tuple{Int64, Any}}, data::MathOptInterface.Nonlinear.Model, expr::MathOptInterface.Nonlinear.Expression, x::Expr, parent_index::Int64)
   @ MathOptInterface.Nonlinear ~/.julia/packages/MathOptInterface/BlCD1/src/Nonlinear/parse.jl:115
 [3] parse_expression(data::MathOptInterface.Nonlinear.Model, expr::MathOptInterface.Nonlinear.Expression, x::Expr, parent_index::Int64)
   @ MathOptInterface.Nonlinear ~/.julia/packages/MathOptInterface/BlCD1/src/Nonlinear/parse.jl:102
 [4] parse_expression
   @ ~/.julia/packages/MathOptInterface/BlCD1/src/Nonlinear/parse.jl:14 [inlined]
 [5] set_objective
   @ ~/.julia/packages/MathOptInterface/BlCD1/src/Nonlinear/model.jl:43 [inlined]
 [6] set_nonlinear_objective(model::Model, sense::MathOptInterface.OptimizationSense, x::Expr)
   @ JuMP ~/.julia/packages/JuMP/H2SWp/src/nlp.jl:167
 [7] macro expansion
   @ ~/.julia/packages/JuMP/H2SWp/src/macros.jl:2721 [inlined]
 [8] top-level scope
   @ ~/Documents/Personnel/ModelPredictiveControl.jl/example/test.jl:13
```
There is no error if I replace `nvar=1` to `nvar=2`. 

Using the debugger, I found that the registered function `:Jfunc` is saved into `optim.nlp_model.operators.univariate_operators` if `nvar=1`, and into `optim.nlp_model.operators.multivariate_operators`, if `nvar > 1`. I think that's the main culprit.

Thanks for all the great works !

Francis

--- a/src/Nonlinear/parse.jl
+++ b/src/Nonlinear/parse.jl
@@ -112,6 +112,9 @@ function _parse_expression(stack, data, expr, x, parent_index)
         if length(x.args) == 2 && !isexpr(x.args[2], :...)
             _parse_univariate_expression(stack, data, expr, x, parent_index)
         else
+            # The call is either n-ary, or it is a splat, in which case we
+            # cannot tell just yet whether the expression is unary or nary.
+            # Punt to multivariate and try to recover later.
             _parse_multivariate_expression(stack, data, expr, x, parent_index)
         end
     elseif isexpr(x, :comparison)
@@ -177,8 +180,15 @@ function _parse_multivariate_expression(
     @assert isexpr(x, :call)
     id = get(data.operators.multivariate_operator_to_id, x.args[1], nothing)
     if id === nothing
-        @assert x.args[1] in data.operators.comparison_operators
-        _parse_inequality_expression(stack, data, expr, x, parent_index)
+        if haskey(data.operators.univariate_operator_to_id, x.args[1])
+            # It may also be a unary variate operator with splatting.
+            _parse_univariate_expression(stack, data, expr, x, parent_index)
+        elseif x.args[1] in data.operators.comparison_operators
+            # Or it may be a binary (in)equality operator.
+            _parse_inequality_expression(stack, data, expr, x, parent_index)
+        else
+            throw(MOI.UnsupportedNonlinearOperator(x.args[1]))
+        end
         return
     end
     push!(expr.nodes, Node(NODE_CALL_MULTIVARIATE, id, parent_index))

--- a/test/Nonlinear/Nonlinear.jl
+++ b/test/Nonlinear/Nonlinear.jl
@@ -1089,7 +1089,7 @@ function test_parse_univariate_splatting()
     MOI.Nonlinear.register_operator(model, :f, 1, x -> 2x)
     x = [MOI.VariableIndex(1)]
     @test MOI.Nonlinear.parse_expression(model, :(f($x...))) ==
-        MOI.Nonlinear.parse_expression(model, :(f($(x[1]))))
+          MOI.Nonlinear.parse_expression(model, :(f($(x[1]))))
     return
 end
 

--- a/test/Nonlinear/Nonlinear.jl
+++ b/test/Nonlinear/Nonlinear.jl
@@ -1084,6 +1084,25 @@ function test_ListOfSupportedNonlinearOperators()
     return
 end
 
+function test_parse_univariate_splatting()
+    model = MOI.Nonlinear.Model()
+    MOI.Nonlinear.register_operator(model, :f, 1, x -> 2x)
+    x = [MOI.VariableIndex(1)]
+    @test MOI.Nonlinear.parse_expression(model, :(f($x...))) ==
+        MOI.Nonlinear.parse_expression(model, :(f($(x[1]))))
+    return
+end
+
+function test_parse_unsupported_operator()
+    model = MOI.Nonlinear.Model()
+    x = [MOI.VariableIndex(1)]
+    @test_throws(
+        MOI.UnsupportedNonlinearOperator(:f),
+        MOI.Nonlinear.parse_expression(model, :(f($x...))),
+    )
+    return
+end
+
 end
 
 TestNonlinear.runtests()

